### DISCUSSION TITLE
[@types/lodash] New version of types breaks type for _.throttle

### DISCUSSION THREAD

**Discussion by kaiyoma** [DefinitelyTyped/DefinitelyTyped#53665] [Issues with a @types package]
Copied from the original issue: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/47571

In version 4.14.160, this code works fine:

```
const throttleValue = _.throttle(_.identity, interval);
const throttledValue = throttleValue(value);
```

In version 4.14.161, this no longer works and incorrect errors about the wrong number of arguments are being reported.

@dobesv made this change here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46971

I started a discussion in the Discord and it was suggested that perhaps the original commit should be reverted and reworked. It sounds like the new approach doesn't work correctly with overloads and/or generics, which is kind of a deal-breaker.

Edit: A workaround for now is to wrap the function to be throttled in a fat-arrow function:

```
const throttleValue = _.throttle(x => _.identity(x), interval);
const throttledValue = throttleValue(value);
```

**Comment by sandersn** [comment]
Couple of notes about the workaround:

1. It might require type annotations on the arrow's parameters.
2. It produces a function with a single signature -- it's not overloaded. I couldn't figure out a way to produce a correctly-typed, overloaded function from the current declaration of throttle.

_Edit: could -> couldn't_

**Reply by kaiyoma** [reply]
"could figure out" = "couldn't figure out"?

**Comment by dobesv** [comment]
Could you clarify the error you are getting?

Is it on the first line or the second?

Is it that the function type returned by throttle doesn't preserve the argument types of the original function (`identity`) properly?

Is the issue about losing overloads, or perhaps losing the generic nature of the function?

Did you try using a cast instead of wrapping with an arrow function, e.g. `_.throttle(_.identity as (<T>(t:T) => T), interval);` ?

**Reply by kaiyoma** [reply]
The original error I got (with the original repro code) is `Expected 0 arguments, but got 1.` (on the first line), which cascades into further errors.

**Reply by kaiyoma** [reply]
> Is it that the function type returned by throttle doesn't preserve the argument types of the original function (identity) properly?

I believe that's the working theory? There's been more discussion in Discord too: https://discord.com/channels/508357248330760243/757992230077333655

**Reply by sandersn** [reply]
Yes, `Parameters` chooses the paramers of the first overload only. It's a limitation of the Typescript compiler.

**Reply by dobesv** [reply]
Ah I see.  Interesting that this worked before, I guess `throttle` just pretended to return the exact same type as its parameter?

**Reply by sandersn** [reply]
That's right. In fact, before the change, the return type was just the input function type intersected with `{ cancel, flush }` I don't think `flush` returned the right type either. The current types *are* more accurate, but they only work with the first overload.
